impl/client/src/main/javascript/web/pentaho/type/action/ComplexChangeset.js

```diff
@@ -233,6 +234,64 @@ define([
       }
     },
 
+    /** @inheritDoc */
+    compose: function(changeset) {
+
+      if(!(changeset instanceof ListChangeset)) {
+        throw error.argInvalidType("changeset", "pentaho.type.action.ComplexChangeset", typeof changeset);
+      }
+
+      if(changeset === this) {
+        throw error.argInvalid("changeset", "Changeset is equal to itself");
+      }
+
+      var target = changeset.target;
+      if(target !== this.target) {
+        throw error.operInvalid("Can not compose changeset that has a different target");
+      }
+
+      var targetVersion = changeset.targetVersion;
+      if(targetVersion < this.targetVersion) {
+        throw error.operInvalid("Can not compose changeset with an outdated target version");
+      }
+
+      var transaction =  changeset.transaction;
+      var resultChangeset = new ComplexChangeset(transaction, target);
+
+      resultChangeset.__targetVersion = targetVersion;
+      resultChangeset.__setReadOnlyInternal();
+
+      // populate result changeset with changes from the previous changeset
```

Comment:
Please format comments like others around — First letter is upper case. Finish with a period.